File: TypeScript/strain.ts
export function keep<T>(arr: T[], pred: (item: T) => boolean): T[] {
    const kept: T[] = [];

    for (const item of arr) {
        if (pred(item)) {
            kept.push(item);
        }
    }

    return kept;
}

export function discard<T>(arr: T[], pred: (item: T) => boolean): T[] {
    const kept: T[] = [];

    for (const item of arr) {
        if (!pred(item)) {
            kept.push(item);
        }
    }

    return kept;
}


const keepFirstAndLast = keep<number>([1, 2, 3], (e: number) => e % 2 === 1);
console.log(keepFirstAndLast); // [1, 3]

const discardFirstAndLast = discard<number>([1, 2, 3], (e: number) => e % 2 === 1);
console.log(discardFirstAndLast); // [2]

File: C#/Strain/Strain.cs
using System;
using System.Collections.Generic;

public static class Strain
{
    public static IEnumerable<T> Keep<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
    { 
        var kept = new List<T>();

        foreach(var item in collection) {
            if (predicate(item)) {
                kept.Add(item);
            }
        }

        return kept;
    }

    public static IEnumerable<T> Discard<T>(this IEnumerable<T> collection, Func<T, bool> predicate)
    {
        var kept = new List<T>();

        foreach (var item in collection)
        {
            if (!predicate(item))
            {
                kept.Add(item);
            }
        }

        return kept;
    }
}